collections/src/main/scala/strawman/collection/immutable/LazyList.scala

```diff
@@ -475,8 +465,26 @@ object LazyList extends LazyListFactory[LazyList] {
       tlEvaluated = true
       tl
     }
-    def force: Evaluated[A] = Some((head, tail))
-    override def toString: String =
+    def force: this.type = {
+      // Use standard 2x 1x iterator trick for cycle detection ("those" is slow one)
+      var these, those: LazyList[A] = this
+
+      def hareStep(): Unit = {
```

Comment:
DRY, I know, but this method causes boxing because it assigns the local variable `these`. So let's just inline it (manually)

```
$ scala -Xprint:cleanup
scala> def f = { var x = "hi"; def step() = { x += "-" }; step(); step(); x }
    def f(): String = {
      var x: runtime.ObjectRef = scala.runtime.ObjectRef.create("hi");
      iw.this.step$1(x);
      iw.this.step$1(x);
      x.elem.$asInstanceOf[String]()
    };
    final private[this] def step$1(x$1: runtime.ObjectRef): Unit = x$1.elem = x$1.elem.$asInstanceOf[String]().+("-");
```

See how `x` is boxed into an `ObjectRef`.

Reply:
Fixed. 
@lrytz please, can you give me a direction to learn more about purpose of `ObjectRef` and why wrapping in the case required?

Reply:
In the JVM there are no nested methods. So if a nested method in Scala accesses a local `val` of the enclosing method, that val gets passed as argument to the lifted nested method:

```
$ scala -Xprint:cleanup
scala> def f = { val x = 1; def g = x; g }
    def f(): Int = {
      val x: Int = 1;
      iw.this.g$1(x)
    };
    final private[this] def g$1(x$1: Int): Int = x$1;
```

Now if a nested method *modifies* a local variable of the enclosing method, passing the variable as argument is not enough, as the lifted method could not modify the local variable of the other method. So the compiler wraps the local variable in a mutable box.

```
scala> def f = { var x = 1; def g = {x = 2}; g; x }
    def f(): Int = {
      var x: runtime.IntRef = scala.runtime.IntRef.create(1);
      iw.this.g$1(x);
      x.elem
    };
    final private[this] def g$1(x$1: runtime.IntRef): Unit = x$1.elem = 2;
```

Reply:
Thx a lot! :)